Python/pylifecycle.c

```diff
@@ -609,7 +610,11 @@ init_interp_create_gil(PyThreadState *tstate, int gil)
 static int
 builtins_dict_watcher(PyDict_WatchEvent event, PyObject *dict, PyObject *key, PyObject *new_value)
 {
-    RARE_EVENT_INC(builtin_dict);
+    PyInterpreterState *interp = _PyInterpreterState_GET();
+    if (event != PyDict_EVENT_CLONED && interp->rare_events.builtin_dict < _Py_MAX_ALLOWED_BUILTINS_MODIFICATIONS) {
```

Comment:
> Maybe I'm just misunderstanding, but shouldn't this be >? As in, if we've modified builtins too many times, we invalidate everything?

Every time the builtins changes any executors that depend on it not changing must be invalidated.
So we  invalidate the executor(s) every time it the builtins changes. 
Upon reaching the threshold we no longer perform those optimizations, so the executors no longer depend on the builtins not changing, and we no longer need to invalidate.

> Also, what is the purpose of excluding PyDict_EVENT_CLONED here?

 Cloning doesn't modify the dict, so we don't care about it.

Reply:
@carljm OOI, why does Cinder care about dict cloning?

Reply:
`PyDict_EVENT_CLONED` is probably a bad name, because grammatically it implies the subject dict was cloned. What it actually means is that the subject dict was empty, and another dict was cloned _into it_, via e.g. `.update()`. So it does represent a change in the subject dict; all dict watcher events do.

I think it is probably irrelevant here, since it is (at least in practice) impossible(?) for builtins or module globals to be empty when code is running.